File: trees_and_graphs/minimum_depth_of_binary_tree.java
public int minDepthRecursive(TreeNode root)
{
    if (root == null)
    {
        return 0;
    }

    if (root.left == null && root.right == null)
    {
        return 1;
    }

    int min_depth = Integer.MAX_VALUE;
    if (root.left != null)
    {
        min_depth = Math.min(minDepthRecursive(root.left), min_depth);
    }

    if (root.right != null)
    {
        min_depth = Math.min(minDepthRecursive(root.right), min_depth);
    }

    return min_depth + 1;
}

class Pair
{
    TreeNode node;
    int depth;

    Pair(TreeNode node, int depth)
    {
        this.node = node;
        this.depth = depth;
    }
}

public int minDepthDFSIterative(TreeNode root)
{
    if (root == null)
    {
        return 0;
    }

    Stack<Pair> stack = new Stack<>();
    stack.push(new Pair(root, 1));

    int min_depth = Integer.MAX_VALUE;
    while (!stack.empty())
    {
        Pair current = stack.pop();
        TreeNode node = current.node;
        int current_depth = current.depth;

        if (node.left == null && node.right == null)
        {
            min_depth = Math.min(min_depth, current_depth);
        }

        if (node.left != null)
        {
            stack.push(new Pair(node.left, current_depth + 1));
        }

        if (node.right != null)
        {
            stack.push(new Pair(node.right, current_depth + 1));
        }
    }

    return min_depth;
}

public int minDepthBFSIterative(TreeNode root)
{
    if (root == null)
    {
        return 0;
    }

    Queue<Pair> queue = new LinkedList<>();
    queue.offer(new Pair(root, 1));

    int current_depth = 0;
    while (!queue.isEmpty())
    {
        Pair pair = queue.poll();
        TreeNode node = pair.node;
        current_depth = pair.depth;

        if (node.left == null && node.right == null)
        {
            break;
        }

        if (node.left != null)
        {
            queue.offer(new Pair(node.left, current_depth + 1));
        }

        if (node.right != null)
        {
            queue.offer(new Pair(node.right, current_depth + 1));
        }
    }
    
    return current_depth;
}

File: trees_and_graphs/minimum_depth_of_binary_tree.js
const minDepthRecursive = (root) => {
    if (!root)
    {
        return 0;
    }

    if (!root.left && !root.right)
    {
        return 1;
    }

    let min_depth = Number.MAX_SAFE_INTEGER;

    if (root.left)
    {
        min_depth = Math.min(minDepthRecursive(root.left), min_depth);
    }

    if (root.right)
    {
        min_depth = Math.min(minDepthRecursive(root.right), min_depth);
    }

    return min_depth + 1;
};

const minDepthDFSIterative = (root) => {
    if (!root)
    {
        return 0;
    }

    let stack = [[root, 1]];
    let min_depth = Number.MAX_SAFE_INTEGER;

    while (stack.length)
    {
        const [node, current_depth] = stack.pop();
        if (!node.left && !node.right)
        {
            min_depth = Math.min(min_depth, current_depth);
        }

        if (node.left)
        {
            stack.push([node.left, current_depth + 1]);
        }

        if (node.right)
        {
            stack.push([node.right, current_depth + 1]);
        }
    }

    return min_depth;
};

const minDepthBFSIterative = (root) => {
    if (!root)
    {
        return 0;
    }

    let queue = [[root, 1]];
    let current_depth = 0;

    while (queue.length)
    {
        const [node, depth] = queue.shift();
        current_depth = depth;
        if (!node.left && !node.right)
        {
            break;
        }

        if (node.left)
        {
            queue.push([node.left, current_depth + 1]);
        }

        if (node.right)
        {
            queue.push([node.right, current_depth + 1]);
        }
    }

    return current_depth;
};